File: release/SmartLibrarySystem_Delivery_20260320/backend-library/src/main/java/com/example/library/entity/ServiceAppointment.java
package com.example.library.entity;

import jakarta.persistence.*;
import lombok.Data;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;

/**
 * Service appointment entity.
 */
@Entity
@Table(name = "service_appointments")
@Data
public class ServiceAppointment {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "appointment_id")
    private Integer appointmentId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "loan_id")
    private Loan loan;

    @Enumerated(EnumType.STRING)
    @Column(name = "service_type", nullable = false)
    private ServiceType serviceType;

    @Column(name = "scheduled_time", nullable = false)
    private LocalDateTime scheduledTime;

    @Enumerated(EnumType.STRING)
    @Column(name = "method")
    private ServiceMethod method = ServiceMethod.COUNTER;

    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    private AppointmentStatus status = AppointmentStatus.PENDING;

    @Column(name = "notes")
    private String notes;

    @Column(name = "return_location", length = 120)
    private String returnLocation;

    @CreationTimestamp
    @Column(name = "create_time", nullable = false, updatable = false)
    private LocalDateTime createTime;

    @UpdateTimestamp
    @Column(name = "update_time", nullable = false)
    private LocalDateTime updateTime;

    public enum ServiceType {
        RETURN_BOOK, PICKUP_BOOK, CONSULTATION
    }

    public enum ServiceMethod {
        COUNTER, SMART_LOCKER
    }

    public enum AppointmentStatus {
        PENDING, COMPLETED, CANCELLED, MISSED
    }
}

File: release/SmartLibrarySystem_Delivery_20260320/front-android/src/services/serviceAppointment.ts
import type {
  ApiServiceAppointmentCreateDto,
  ApiServiceAppointmentDto,
  PageResponse,
} from "../types/api";
import { request } from "./http";

export interface ServiceAppointment {
  appointmentId: number;
  userId: number;
  username: string;
  userFullName?: string;
  loanId?: number;
  bookTitle?: string;
  serviceType: "RETURN_BOOK" | "PICKUP_BOOK" | "CONSULTATION";
  method: "COUNTER" | "SMART_LOCKER";
  status: "PENDING" | "COMPLETED" | "CANCELLED" | "MISSED";
  scheduledTime: string;
  returnLocation?: string;
  notes?: string;
  createTime: string;
  updateTime?: string;
}

function mapAppointment(dto: ApiServiceAppointmentDto): ServiceAppointment {
  return {
    appointmentId: dto.appointmentId,
    userId: dto.userId,
    username: dto.username,
    userFullName: dto.userFullName,
    loanId: dto.loanId,
    bookTitle: dto.bookTitle,
    serviceType: dto.serviceType,
    method: dto.method,
    status: dto.status,
    scheduledTime: dto.scheduledTime,
    returnLocation: dto.returnLocation,
    notes: dto.notes,
    createTime: dto.createTime,
    updateTime: dto.updateTime,
  };
}

export const serviceAppointmentService = {
  async createAppointment(
    payload: ApiServiceAppointmentCreateDto,
  ): Promise<ServiceAppointment> {
    const response = await request<ApiServiceAppointmentDto, ApiServiceAppointmentCreateDto>({
      url: "/service-appointments",
      method: "POST",
      data: payload,
      auth: true,
    });

    return mapAppointment(response);
  },

  async getMyAppointments(page = 0, size = 20): Promise<PageResponse<ServiceAppointment>> {
    const response = await request<PageResponse<ApiServiceAppointmentDto>>({
      url: "/service-appointments/me",
      query: { page, size },
      auth: true,
    });

    return {
      ...response,
      content: (response.content ?? []).map(mapAppointment),
    };
  },

  async cancelAppointment(appointmentId: number): Promise<void> {
    await request<void>({
      url: `/service-appointments/${appointmentId}/cancel`,
      method: "PUT",
      auth: true,
    });
  },
};
